README.md
# node-percipio-entitlementsreport

Retrieve [License Consumption Report](https://documentation.skillsoft.com/en_us/percipio/Content/A_Administrator/admn_rpt_entitlements.htm) data from Percipio and save locally.

## Requirements

1. A Skillsoft [Percipio](https://www.skillsoft.com/platform-solution/percipio/) Site
1. A [Percipio Service Account](https://documentation.skillsoft.com/en_us/pes/3_services/service_accounts/pes_service_accounts.htm) with permission for accessing [REPORTING API](https://documentation.skillsoft.com/en_us/pes/2_understanding_percipio/rest_api/pes_rest_api.htm)

## Environment Configuration

Once you have copied this repository set the following NODE ENV variables, or config the [.env](.env) file

| ENV       | Required | Description                                                                                                                                                                                                                                                                                      |
| --------- | -------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ |
| ORGID     | Required | This is the Percipio Organiation UUID for your Percipio Site                                                                                                                                                                                                                                     |
| BEARER    | Required | This is the Percipio Bearer token for a Service Account with permissions for services.                                                                                                                                                                                         |
| BASEURL | Required | This is set to the base URL for the Percipio data center. For US hosted use: https://api.percipio.com For EU hosted use: https://dew1-api.percipio.com |
| TIMEFRAME | Optional | This is a filter criteria that specifies the timeframe for the results.<br/><br/>The report start/end dates are calculated dynamically based on when the report is submitted date.<br/><br/>Options are: DAY, WEEK, THIRTY_DAYS, CALENDAR_MONTH<br/><br/>If left empty/null THIRTY_DAYS is used. |
| START     | Optional | This is a filter criteria that specifies the START date for the report in ISO8601 format.<br/><br/>The END option must be specified if using this.<br/><br/>The TIMEFRAME option must be null if using this.                                                                                     |
| END     | Optional | This is a filter criteria that specifies the END date for the report in ISO8601 format.<br/><br/>The START option must be specified if using this.<br/><br/>The TIMEFRAME option must be null if using this                                                                                       |

## Configuring the API call
Make any additional config changes in [config/default.js](config/default.js) file, to specify the request criteria for the report other than date ranges.

## How to use it

Run the app

```bash
npm start
```

The Percipio [https://api.percipio.com/reporting/api-docs/#/%2Fv1/requestEntitlementReport](https://api.percipio.com/reporting/api-docs/#/%2Fv1/requestEntitlementReport) API wil be called to generate the report.

The Percipio[https://api.percipio.com/reporting/api-docs/#/%2Fv1/getReportRequest](https://api.percipio.com/reporting/api-docs/#/%2Fv1/getReportRequest) API will then be called to download the generated data.

The default configuration returns JSON and the data will be stored in:

```
results/YYYYMMDD_hhmmss_results.json
```

The timestamp component is based on UTC time when the script runs:

| DATEPART | COMMENTS                            |
| -------- | ----------------------------------- |
| YYYY     | Year (i.e. 1970 1971 ... 2029 2030) |
| MM       | Month Number (i.e. 01 02 ... 11 12) |
| DD       | Day (i.e. 01 02 ... 30 31)          |
| HH       | Hour (i.e. 00 01 ... 22 23)         |
| mm       | Minutes (i.e. 00 01 ... 58 59)      |
| ss       | Seconds (i.e. 00 01 ... 58 59)      |

## Changelog

Please see [CHANGELOG](CHANGELOG.md) for more information what has changed recently.

## License

MIT © martinholden-skillsoft


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

Generated by [`auto-changelog`](https://github.com/CookPete/auto-changelog).

## [v1.2.4](https://github.com/martinholden-skillsoft/node-percipio-entitlementsreport/compare/v1.2.3...v1.2.4) - 2021-01-11

### Commits

- Updated Dependencies [`a6b75bf`](https://github.com/martinholden-skillsoft/node-percipio-entitlementsreport/commit/a6b75bf38830815c670d3274f62235fb13126a70)
- ESLINT fixes [`21d9ada`](https://github.com/martinholden-skillsoft/node-percipio-entitlementsreport/commit/21d9adaed2164ecbf06a55ed0ab68bde1e756e63)
- ESLINT fixes [`8861a78`](https://github.com/martinholden-skillsoft/node-percipio-entitlementsreport/commit/8861a7886fb272a34ff7bc65e97447c88a167e99)

## [v1.2.3](https://github.com/martinholden-skillsoft/node-percipio-entitlementsreport/compare/v1.2.2...v1.2.3) - 2020-11-04

### Commits

- changelog [`c4fcac9`](https://github.com/martinholden-skillsoft/node-percipio-entitlementsreport/commit/c4fcac9dac239bd228adf5f5ec20602cde03649a)
- Updated README [`8d1817b`](https://github.com/martinholden-skillsoft/node-percipio-entitlementsreport/commit/8d1817b703d445112a2f4723d8641278633298b6)
- Fix .env config issue [`a4135a3`](https://github.com/martinholden-skillsoft/node-percipio-entitlementsreport/commit/a4135a343d007e60447276d281da18faee3f5763)

## [v1.2.2](https://github.com/martinholden-skillsoft/node-percipio-entitlementsreport/compare/v1.2.1...v1.2.2) - 2020-11-02

### Commits

- Removed extraneous file [`98dd922`](https://github.com/martinholden-skillsoft/node-percipio-entitlementsreport/commit/98dd922d166e422b409f6a964673d6bae7bb962f)
- changelog [`2ccb382`](https://github.com/martinholden-skillsoft/node-percipio-entitlementsreport/commit/2ccb382fe21d5bb1da0ead55318986268ede9f0e)

## [v1.2.1](https://github.com/martinholden-skillsoft/node-percipio-entitlementsreport/compare/v1.2.0...v1.2.1) - 2020-11-02

### Commits

- Enable JSON and CSV response handling [`8710ce8`](https://github.com/martinholden-skillsoft/node-percipio-entitlementsreport/commit/8710ce8354583129fa08eafe35abb64688b9c8f7)
- changelog [`82573e2`](https://github.com/martinholden-skillsoft/node-percipio-entitlementsreport/commit/82573e234dd65902f8adb27870eb8202174c8dfb)

## [v1.2.0](https://github.com/martinholden-skillsoft/node-percipio-entitlementsreport/compare/v1.1.0...v1.2.0) - 2020-11-02

### Commits

- Changes to use config and dotenv-safe [`a226585`](https://github.com/martinholden-skillsoft/node-percipio-entitlementsreport/commit/a226585c6d799e3991af1863fb502a7c1db84812)
- Updated README [`e9693c4`](https://github.com/martinholden-skillsoft/node-percipio-entitlementsreport/commit/e9693c4cdd3a5a4e4802bc08aa4b5776f14b6995)
- changelog [`7af8af9`](https://github.com/martinholden-skillsoft/node-percipio-entitlementsreport/commit/7af8af9105e47f74d13ce199e39b0aaf3ff9cc22)

## [v1.1.0](https://github.com/martinholden-skillsoft/node-percipio-entitlementsreport/compare/v1.0.0...v1.1.0) - 2020-10-22

### Commits

- changelog [`8c3a0a3`](https://github.com/martinholden-skillsoft/node-percipio-entitlementsreport/commit/8c3a0a3d380d56deaaa53cca3e73ffa2d893cb4b)
- Fixed Bad Polling Retry Options [`e264220`](https://github.com/martinholden-skillsoft/node-percipio-entitlementsreport/commit/e26422015b84f08975ce4837da34e5c7bbc03eaf)
- Fxied typo [`b46340b`](https://github.com/martinholden-skillsoft/node-percipio-entitlementsreport/commit/b46340b78e6266fe5ef0e77a6d3240d6b661ec25)

## v1.0.0 - 2020-10-22

### Commits

- Initial commit [`e30db48`](https://github.com/martinholden-skillsoft/node-percipio-entitlementsreport/commit/e30db48abcf3ab68b607ec9969cc8f51b5078467)


app.js
require('dotenv-safe').config();

const config = require('config');
const axios = require('axios');
const fs = require('fs');
const Path = require('path');
const _ = require('lodash');
const mkdirp = require('mkdirp');
const promiseRetry = require('promise-retry');
const stringifySafe = require('json-stringify-safe');

const { transports } = require('winston');
const logger = require('./lib/logger');

const pjson = require('./package.json');

const NODE_ENV = process.env.NODE_ENV || 'production';

/**
 * Process the template string and replace the path values
 *
 * @param {string} templateString
 * @param {*} templateVars
 * @return {string}
 */
const processTemplate = (templateString, templateVars) => {
  const compiled = _.template(templateString.replace(/{/g, '${'));
  return compiled(templateVars);
};

/**
 * Call Percipio API
 *
 * @param {*} options
 * @returns
 */
const callPercipio = async (options) =>
  promiseRetry(async (retry, numberOfRetries) => {
    const loggingOptions = {
      label: 'callPercipio',
    };

    const requestUri = processTemplate(options.request.uritemplate, options.request.path);
    options.logger.debug(`Request URI: ${requestUri}`, loggingOptions);

    let requestParams = options.request.query || {};
    requestParams = _.omitBy(requestParams, _.isNil);
    options.logger.debug(
      `Request Querystring Parameters: ${stringifySafe(requestParams)}`,
      loggingOptions
    );

    let requestBody = options.request.body || {};
    requestBody = _.omitBy(requestBody, _.isNil);
    options.logger.debug(`Request Body: ${stringifySafe(requestBody)}`, loggingOptions);

    const axiosConfig = {
      baseURL: options.request.baseURL,
      url: requestUri,
      headers: {
        Authorization: `Bearer ${options.request.bearer}`,
      },
      method: options.request.method,
      timeout: options.request.timeout || 2000,
    };

    if (!_.isEmpty(requestBody)) {
      axiosConfig.data = requestBody;
    }

    if (!_.isEmpty(requestParams)) {
      axiosConfig.params = requestParams;
    }

    options.logger.debug(`Axios Config: ${stringifySafe(axiosConfig)}`, loggingOptions);

    try {
      const response = await axios.request(axiosConfig);
      options.logger.debug(`Response Headers: ${stringifySafe(response.headers)}`, loggingOptions);
      return response;
    } catch (err) {
      options.logger.warn(
        `Trying to get report. Got Error after Attempt# ${numberOfRetries} : ${err}`,
        loggingOptions
      );
      if (err.response) {
        options.logger.debug(
          `Response Headers: ${stringifySafe(err.response.headers)}`,
          loggingOptions
        );
        options.logger.debug(`Response Body: ${stringifySafe(err.response.data)}`, loggingOptions);
      } else {
        options.logger.debug('No Response Object available', loggingOptions);
      }
      if (numberOfRetries < options.retry_options.retries + 1) {
        retry(err);
      } else {
        options.logger.error('Failed to call Percipio', loggingOptions);
      }
      throw err;
    }
  }, options.retry_options);

/**
 * Submit the report request
 *
 * @param {*} options
 * @returns
 */
const submitReport = async (options) => {
  const loggingOptions = {
    label: 'submitReport',
  };
  const opts = options;
  opts.request = opts.reportrequest;
  options.logger.info('Submitting Report Request', loggingOptions);
  return callPercipio(opts);
};

/**
 * Poll for the specified reportRequestId, the polling uses the promiseRetry
 * confguration defined in options.polling_options to determine number of
 * attempts and delays
 *
 * @param {*} options
 * @param {*} reportRequestId
 * @returns
 */
const pollForReport = async (options, reportRequestId) =>
  promiseRetry(async (retry, numberOfRetries) => {
    const loggingOptions = {
      label: 'pollForReport',
    };

    const opts = options;
    opts.request = opts.pollrequest;
    opts.request.path.reportRequestId = reportRequestId;
    opts.request.uri = `${opts.request.baseuri}/reporting/v1/organizations/${opts.request.path.orgId}/report-requests/${opts.request.path.reportRequestId}`;

    options.logger.info('Submitting Report Poll Requests', loggingOptions);
    try {
      const response = await callPercipio(opts);
      if (_.isUndefined(response.data.status)) {
        return response;
      }
      const error = new Error(`Report ${response.data.reportId} status is ${response.data.status}`);
      error.response = response;
      throw error;
    } catch (err) {
      options.logger.warn(
        `Trying to get report. Got Error after Attempt# ${numberOfRetries} of ${options.polling_options.retries} : ${err}`,
        loggingOptions
      );
      if (err.response) {
        options.logger.debug(
          `Response Headers: ${stringifySafe(err.response.headers)}`,
          loggingOptions
        );
        options.logger.debug(`Response Body: ${stringifySafe(err.response.data)}`, loggingOptions);
      } else {
        options.logger.debug('No Response Object available', loggingOptions);
      }
      if (numberOfRetries < options.polling_options.retries + 1) {
        retry(err);
      } else {
        options.logger.error('Failed to retrieve report', loggingOptions);
      }
      throw err;
    }
  }, options.polling_options);

/**
 * Call the API to generate the report, and the poll for it
 * pass thru a stream and return the path to the file
 *
 * @param {*} options
 * @returns {string} file path
 */
const getAllReportDataAndSave = async (options) =>
  // eslint-disable-next-line no-async-promise-executor
  new Promise(async (resolve, reject) => {
    const loggingOptions = {
      label: 'getAllReportDataAndSave',
    };

    const opts = options;

    const filename = opts.output.filename || 'result';

    const outputFile = Path.join(opts.output.path, filename);

    try {
      submitReport(options)
        .then((submitResponse) => {
          options.logger.info(`Report Id: ${submitResponse.data.id}`, loggingOptions);
          pollForReport(options, submitResponse.data.id)
            .then((reportResponse) => {
              // Handle a JSON response
              if (_.isObject(reportResponse.data) && !_.isString(reportResponse.data)) {
                options.logger.info(
                  `Records Downloaded ${reportResponse.data.length.toLocaleString()}`,
                  loggingOptions
                );
                const outputStream = fs.createWriteStream(outputFile);

                outputStream.on('finish', () => {
                  options.logger.info(`Records Saved. Path: ${outputFile}`, loggingOptions);
                  resolve(outputFile);
                });

                outputStream.write(stringifySafe(reportResponse.data, null, 2));
                outputStream.end();
              } else if (!_.isObject(reportResponse.data) && _.isString(reportResponse.data)) {
                // Handle a CSV response
                options.logger.info(
                  `Records Downloaded. Download Size: ${reportResponse.data.length.toLocaleString()} bytes`,
                  loggingOptions
                );
                const outputStream = fs.createWriteStream(outputFile);

                outputStream.on('finish', () => {
                  options.logger.info(`Records Saved. Path: ${outputFile}`, loggingOptions);
                  resolve(outputFile);
                });

                outputStream.write(reportResponse.data);
                outputStream.end();
              } else {
                options.logger.warn(
                  'Response is not valid JSON or CSV. No results file created.',
                  loggingOptions
                );
                resolve(null);
              }
            })
            .catch((err) => {
              options.logger.error(`Error:  ${err}`, loggingOptions);
            });
        })
        .catch((err) => {
          options.logger.error(`Error:  ${err}`, loggingOptions);
        });
    } catch (err) {
      options.logger.error('ERROR: trying to generate report results', loggingOptions);
      reject(err);
    }
  });
/**
 * Process the Percipio call
 *
 * @param {*} options
 * @returns
 */
const main = async (configOptions) => {
  const loggingOptions = {
    label: 'main',
  };

  const options = configOptions || null;

  options.logger = logger;

  if (_.isNull(options)) {
    options.logger.error('Invalid configuration', loggingOptions);
    return false;
  }

  // Set logging to silly level for dev
  if (NODE_ENV.toUpperCase() === 'DEVELOPMENT') {
    options.logger.level = 'debug';
  } else {
    options.logger.level = options.debug.loggingLevel;
  }

  // Create logging folder if one does not exist
  if (!_.isNull(options.debug.path)) {
    if (!fs.existsSync(options.debug.path)) {
      mkdirp(options.debug.path);
    }
  }

  // Create outpur folder if one does not exist
  if (!_.isNull(options.output.path)) {
    if (!fs.existsSync(options.output.path)) {
      mkdirp(options.output.path);
    }
  }

  // Add logging to a file
  options.logger.add(
    new transports.File({
      filename: Path.join(options.debug.path, options.debug.filename),
      options: {
        flags: 'w',
      },
    })
  );

  options.logger.info(`Start ${pjson.name} - v${pjson.version}`, loggingOptions);

  options.logger.debug(`Options: ${stringifySafe(options)}`, loggingOptions);

  options.logger.info('Calling Percipio', loggingOptions);
  await getAllReportDataAndSave(options).catch((err) => {
    options.logger.error(`Error:  ${err}`, loggingOptions);
  });
  options.logger.info(`End ${pjson.name} - v${pjson.version}`, loggingOptions);
  return true;
};

try {
  main(config);
} catch (error) {
  throw new Error(`An error occured during configuration. ${error.message}`);
}


config/config.global.js
const moment = require('moment');

const config = {};

// Indicates a name for the configuration
config.customer = 'none';
config.startTimestamp = moment().utc().format('YYYYMMDD_HHmmss');

// DEBUG Options - Enables the check for Fiddler, if running the traffic is routed thru Fiddler
config.debug = {};
// Debug logging
// One of the supported default logging levels for winston - see https://github.com/winstonjs/winston#logging-levels
config.debug.loggingLevel = 'info';
config.debug.path = 'logs';
config.debug.filename = `app_${config.startTimestamp}.log`;

config.output = {};
// Output path
config.output.path = 'results/output';
// Filename
config.output.filename = `${config.startTimestamp}_results.json`;

// Report Generation Request
config.reportrequest = {};
// Request Timeout
config.reportrequest.timeout = 2000;
// Bearer Token
config.reportrequest.bearer = null;
// Base URI to Percipio API
config.reportrequest.baseuri = process.env.EUDC
  ? 'https://dew1-api.percipio.com'
  : 'https://api.percipio.com';
// Request Path Parameters
config.reportrequest.path = {};
/**
 * Name: orgId
 * Description: Organization UUID
 * Required: true
 * Type: string
 * Format: uuid
 */
config.reportrequest.path.orgId = null;
// Request Query string Parameters
config.reportrequest.query = {};
// Request Body
config.reportrequest.body = null;
// Method
config.reportrequest.method = 'post';
// The Service Path
config.reportrequest.uri = `${config.reportrequest.baseuri}/reporting/v1/organizations/${config.reportrequest.path.orgId}/report-requests/entitlements`;

// Request
config.pollrequest = {};
// Request Timeout
config.pollrequest.timeout = 2000;
// Bearer Token
config.pollrequest.bearer = null;
// Base URI to Percipio API
config.pollrequest.baseuri = process.env.EUDC
  ? 'https://dew1-api.percipio.com'
  : 'https://api.percipio.com';
// Request Path Parameters
config.pollrequest.path = {};
/**
 * Name: orgId
 * Description: Organization UUID
 * Required: true
 * Type: string
 * Format: uuid
 */
config.pollrequest.path.orgId = null;
// Request Query string Parameters
config.pollrequest.query = {};
// Request Body
config.pollrequest.body = null;
// Method
config.pollrequest.method = 'get';
// The Service Path
config.pollrequest.uri = `${config.pollrequest.baseuri}/reporting/v1/organizations/${config.pollrequest.path.orgId}/report-requests/${config.pollrequest.path.reportRequestId}`;

// Global Web Retry Options for promise retry
// see https://github.com/IndigoUnited/node-promise-retry#readme
config.retry_options = {};
config.retry_options.retries = 3;
config.retry_options.minTimeout = 1000;
config.retry_options.maxTimeout = 2000;

/*
Polling options for retrying report availability
see https://github.com/IndigoUnited/node-promise-retry#readme
options is a JS object that can contain any of the following keys:
retries: The maximum amount of times to retry the operation.Default is 10.
Seting this to 1 means do it once, then retry it once.
factor: The exponential factor to use.Default is 2.
minTimeout: The number of milliseconds before starting the first retry.Default is 1000.
maxTimeout: The maximum number of milliseconds between two retries.Default is Infinity.
randomize: Randomizes the timeouts by multiplying with a factor between 1 to 2. Default is false.
*/

config.polling_options = {};
config.polling_options.retries = 20;
config.polling_options.minTimeout = 60 * 1000;
config.polling_options.maxTimeout = Infinity;
// Using Factor=1 means polling at fixed interval of minTimeout
config.polling_options.factor = 1;

module.exports = config;


config/config.default.js
const config = require('./config.global');

config.customer = 'default';

// Debug logging
// One of the supported default logging levels for winston - see https://github.com/winstonjs/winston#logging-levels
// config.debug.loggingLevel = 'debug';
config.debug.path = 'results/output';
config.debug.filename = `${config.customer}_${config.startTimestamp}.log`;

// Output path
config.output.path = 'results/output';
config.output.filename = `${config.customer}_entitlements_${config.startTimestamp}.json`;

// Report Generation Request
// Request
config.reportrequest = {};
// Timeout
config.reportrequest.timeout = 2000;

// Bearer Token
config.reportrequest.bearer = process.env.BEARER || null;
// Base URI to Percipio API
config.reportrequest.baseuri = process.env.EUDC
  ? 'https://dew1-api.percipio.com'
  : 'https://api.percipio.com';
// Request Path Parameters
config.reportrequest.path = {};
/**
 * Name: orgId
 * Description : Organization UUID
 * Required: true
 * Type: string
 * Format: uuid
 */
config.reportrequest.path.orgId = process.env.ORGID || null;

// Request Query string Parameters
config.reportrequest.query = {};

// Request Body
config.reportrequest.body = {};
/**
 * Name: start
 * Description : Start date for events retrieval
 * Type: string
 * Format: date-time
 * Example: 2018-01-01T10:10:24Z
 */
config.reportrequest.body.start = process.env.START || null;
/**
 * Name: end
 * Description : End date for events retrieval
 * Type: string
 * Format: date-time
 * Example: 2018-01-10T10:20:24Z
 */
config.reportrequest.body.end = process.env.END || null;
/**
 * Name: timeFrame
 * Description : To calculate the start/end date dynamically based on timeframe and when the
 * report is submitted date. [User can submit an absolute date range (start/end date) or a
 * relative date range (timeframe) but never both]
 * Type: string
 * Enum: DAY, WEEK, THIRTY_DAYS, CALENDAR_MONTH
 */
config.reportrequest.body.timeFrame = config.reportrequest.body.start
  ? null
  : process.env.TIMEFRAME || 'THIRTY_DAYS';
/**
 * Name: audience
 * Description : Audience filter, defaults to all
 * Type: string
 */
config.reportrequest.body.audience = 'ALL';
/**
 * Name: user
 * Description : Users filter, comma delimited and defaults to all
 * Type: string
 */
config.reportrequest.body.user = 'ALL';
/**
 * Name: collection
 * Description : Collection title filter, comma delimited and defaults to all
 * Type: string
 */
config.reportrequest.body.collection = 'ALL';
/**
 * Name: sort
 * Description : The field to use for sorting and which order to sort in, if sort is not
 * included the results will be returned ascending by collection
 * Type: object
 */
config.reportrequest.body.sort = {};

/**
 * Name: field
 * Type: string
 * Enum: collection, userId, firstName, lastName, emailAddress, entitledByAudience, consumed
 */
config.reportrequest.body.sort.field = 'collection';
/**
 * Name: order
 * Type: string
 * Enum: asc, desc
 */
config.reportrequest.body.sort.order = 'asc';
/**
 * Name: sftpId
 * Description : SFTP Id associated with OrgId
 * Type: string
 * Format: uuid
 */
config.reportrequest.body.sftpId = null;
/**
 * Name: isFileRequiredInSftp
 * Description : Generated files are required to deliver in the respected sftp location.
 * Default value is true.
 * Type: boolean
 */
config.reportrequest.body.isFileRequiredInSftp = false;
/**
 * Name: zip
 * Description : Generate the reports in zip file format. Default value is false.
 * Type: boolean
 */
config.reportrequest.body.zip = null;
/**
 * Name: encrypt
 * Description : Generate the report file as PGP encrypted file. Default value is false.
 * Type: boolean
 */
config.reportrequest.body.encrypt = null;
/**
 * Name: formatType
 * Description : Format Type, defaults to JSON, the value is extracted from Accept attribute
 * in header
 * Type: string
 * Enum: JSON, CSV, TXT
 */
config.reportrequest.body.formatType = 'JSON';
/**
 * Name: fileMask
 * Description : Absolute or masked pattern for the generated report file. Example file masks
 * - fileName_{DD}{MM}{YYYY}, fileName_{ORG_ID}
 * Type: string
 */
config.reportrequest.body.fileMask = null;
/**
 * Name: folderName
 * Description : custom folder under sftp reports wherein the generated report file is to be
 * placed.
 * Type: string
 */
config.reportrequest.body.folderName = null;
/**
 * Name: includeMillisInFilename
 * Description : Generate files with unix based timestamp. Example -
 * fileName.csv.1561642446608
 * Type: boolean
 */
config.reportrequest.body.includeMillisInFilename = null;

// Method
config.reportrequest.method = 'post';
// The Service Path
config.reportrequest.uri = `${config.reportrequest.baseuri}/reporting/v1/organizations/${config.reportrequest.path.orgId}/report-requests/entitlements`;

// Report Polling Request
config.pollrequest = {};
// Request Timeout
config.pollrequest.timeout = 2000;
// Bearer Token
config.pollrequest.bearer = process.env.BEARER || null;
// Base URI to Percipio API
config.pollrequest.baseuri = process.env.EUDC
  ? 'https://dew1-api.percipio.com'
  : 'https://api.percipio.com';
// Request Path Parameters
config.pollrequest.path = {};
/**
 * Name: orgId
 * Description : Organization UUID
 * Required: true
 * Type: string
 * Format: uuid
 */
config.pollrequest.path.orgId = process.env.ORGID || null;
/**
 * Name: reportRequestId
 * Description : Handle to access the report content
 * Required: true
 * Type: string
 */
config.pollrequest.path.reportRequestId = null;

// Request Query string Parameters
config.pollrequest.query = {};

// Request Body
config.pollrequest.body = null;

// Method
config.pollrequest.method = 'get';
// The Service Path
config.pollrequest.uri = `${config.pollrequest.baseuri}/reporting/v1/organizations/${config.pollrequest.path.orgId}/report-requests/${config.pollrequest.path.reportRequestId}`;

module.exports = config;


config/default.js
const moment = require('moment');
const defer = require('config/defer').deferConfig;

const config = {};

// Indicates a name for the configuration
config.customer = 'default';
config.startTimestamp = moment().utc().format('YYYYMMDD_HHmmss');
// Null logger object
config.logger = {};

// DEBUG Options - Enables the check for Fiddler, if running the traffic is routed thru Fiddler
config.debug = {};
// Debug logging
// One of the supported default logging levels for winston - see https://github.com/winstonjs/winston#logging-levels
config.debug.loggingLevel = 'info';
config.debug.path = 'results';
config.debug.filename = defer((cfg) => `${cfg.startTimestamp}_results.log`);

config.output = {};
// Output path
config.output.path = 'results';
// Filename
config.output.filename = defer((cfg) => `${cfg.startTimestamp}_results.json`);

// Report Generation Request
config.reportrequest = {};
// Request Timeout
config.reportrequest.timeout = 2000;
// Bearer Token
config.reportrequest.bearer = null;
// Base URI to Percipio API
config.reportrequest.baseURL = null;
// Request Path Parameters
config.reportrequest.path = {};
/**
 * Name: orgId
 * Description: Organization UUID
 * Required: true
 * Type: string
 * Format: uuid
 */
config.reportrequest.path.orgId = null;

// Request Query string Parameters
config.reportrequest.query = {};

// Request Body
config.reportrequest.body = {};
/**
 * Name: start
 * Description : Start date for events retrieval
 * Type: string
 * Format: date-time
 * Example: 2018-01-01T10:10:24Z
 */
config.reportrequest.body.start = process.env.START || null;
/**
 * Name: end
 * Description : End date for events retrieval
 * Type: string
 * Format: date-time
 * Example: 2018-01-10T10:20:24Z
 */
config.reportrequest.body.end = process.env.END || null;
/**
 * Name: timeFrame
 * Description : To calculate the start/end date dynamically based on timeframe and when the
 * report is submitted date. [User can submit an absolute date range (start/end date) or a
 * relative date range (timeframe) but never both]
 * Type: string
 * Enum: DAY, WEEK, THIRTY_DAYS, CALENDAR_MONTH
 */
config.reportrequest.body.timeFrame = config.reportrequest.body.start
  ? null
  : process.env.TIMEFRAME || 'THIRTY_DAYS';
/**
 * Name: audience
 * Description : Audience filter, defaults to all
 * Type: string
 */
config.reportrequest.body.audience = 'ALL';
/**
 * Name: user
 * Description : Users filter, comma delimited and defaults to all
 * Type: string
 */
config.reportrequest.body.user = 'ALL';
/**
 * Name: collection
 * Description : Collection title filter, comma delimited and defaults to all
 * Type: string
 */
config.reportrequest.body.collection = 'ALL';
/**
 * Name: sort
 * Description : The field to use for sorting and which order to sort in, if sort is not
 * included the results will be returned ascending by collection
 * Type: object
 */
config.reportrequest.body.sort = {};

/**
 * Name: field
 * Type: string
 * Enum: collection, userId, firstName, lastName, emailAddress, entitledByAudience, consumed
 */
config.reportrequest.body.sort.field = 'collection';
/**
 * Name: order
 * Type: string
 * Enum: asc, desc
 */
config.reportrequest.body.sort.order = 'asc';
/**
 * Name: sftpId
 * Description : SFTP Id associated with OrgId
 * Type: string
 * Format: uuid
 */
config.reportrequest.body.sftpId = null;
/**
 * Name: isFileRequiredInSftp
 * Description : Generated files are required to deliver in the respected sftp location.
 * Default value is true.
 * Type: boolean
 */
config.reportrequest.body.isFileRequiredInSftp = false;
/**
 * Name: zip
 * Description : Generate the reports in zip file format. Default value is false.
 * Type: boolean
 */
config.reportrequest.body.zip = null;
/**
 * Name: encrypt
 * Description : Generate the report file as PGP encrypted file. Default value is false.
 * Type: boolean
 */
config.reportrequest.body.encrypt = null;
/**
 * Name: formatType
 * Description : Format Type, defaults to JSON, the value is extracted from Accept attribute
 * in header
 * Type: string
 * Enum: JSON, CSV, TXT
 */
config.reportrequest.body.formatType = 'JSON';
/**
 * Name: fileMask
 * Description : Absolute or masked pattern for the generated report file. Example file masks
 * - fileName_{DD}{MM}{YYYY}, fileName_{ORG_ID}
 * Type: string
 */
config.reportrequest.body.fileMask = null;
/**
 * Name: folderName
 * Description : custom folder under sftp reports wherein the generated report file is to be
 * placed.
 * Type: string
 */
config.reportrequest.body.folderName = null;
/**
 * Name: includeMillisInFilename
 * Description : Generate files with unix based timestamp. Example -
 * fileName.csv.1561642446608
 * Type: boolean
 */
config.reportrequest.body.includeMillisInFilename = null;

// Method
config.reportrequest.method = 'post';
// The Service Path the placeholders {} are replaced by values from *.path
config.reportrequest.uritemplate =
  '/reporting/v1/organizations/{orgId}/report-requests/entitlements';

// Request
config.pollrequest = {};
// Request Timeout
config.pollrequest.timeout = 2000;
// Bearer Token
config.pollrequest.bearer = null;
// Base URI to Percipio API
config.pollrequest.baseURL = null;
// Request Path Parameters
config.pollrequest.path = {};
/**
 * Name: orgId
 * Description: Organization UUID
 * Required: true
 * Type: string
 * Format: uuid
 */
config.pollrequest.path.orgId = null;
/**
 * Name: reportRequestId
 * Description : Handle to access the report content
 * Required: true
 * Type: string
 */
config.pollrequest.path.reportRequestId = null;
// Request Query string Parameters
config.pollrequest.query = {};
// Request Body
config.pollrequest.body = null;
// Method
config.pollrequest.method = 'get';
// The Service Path the placeholders {} are replaced by values from *.path
config.pollrequest.uritemplate =
  '/reporting/v1/organizations/{orgId}/report-requests/{reportRequestId}';

// Global Web Retry Options for promise retry
// see https://github.com/IndigoUnited/node-promise-retry#readme
config.retry_options = {};
config.retry_options.retries = 3;
config.retry_options.minTimeout = 1000;
config.retry_options.maxTimeout = 2000;

/*
Polling options for retrying report availability
see https://github.com/IndigoUnited/node-promise-retry#readme
options is a JS object that can contain any of the following keys:
retries: The maximum amount of times to retry the operation.Default is 10.
Seting this to 1 means do it once, then retry it once.
factor: The exponential factor to use.Default is 2.
minTimeout: The number of milliseconds before starting the first retry.Default is 1000.
maxTimeout: The maximum number of milliseconds between two retries.Default is Infinity.
randomize: Randomizes the timeouts by multiplying with a factor between 1 to 2. Default is false.
*/

config.polling_options = {};
config.polling_options.retries = 20;
config.polling_options.minTimeout = 60 * 1000;
config.polling_options.maxTimeout = Infinity;
// Using Factor=1 means polling at fixed interval of minTimeout
config.polling_options.factor = 1;

module.exports = config;
